Tests don't build

Any non-trivial test in `ykrt/src/mt.rs` such as:

```rust
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hot_threshold_passed() {
        let hot_thrsh: HotThreshold = 1500;
        let mt = MT::global();
        let loc = Location::new();
        for i in 0..hot_thrsh {
            MT::transition_location(&loc);
        }
    }
}
```

causes a linker error:

```
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-m64" "/home/ltratt/yk/target/debug/deps/ykrt-e549a82218bf7b47.13mjq23so3byahxg.rcgu.o"
<vast swathes of stuff cut>
          /home/ltratt/yk/ykllvmwrap/src/ykllvmwrap.cc:186: undefined reference to `__ykutil_get_llvmbc_section'
          collect2: error: ld returned 1 exit status
          
  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

warning: `ykrt` (lib test) generated 2 warnings
error: could not compile `ykrt` due to previous error; 2 warnings emitted
```

I wonder if this is something to do with the `thread_local!` we use?

--- a/ykcapi/Cargo.toml
+++ b/ykcapi/Cargo.toml
@@ -13,7 +13,6 @@ libc = "0.2.97"
 ykllvmwrap = { path = "../ykllvmwrap" }
 ykrt = { path = "../ykrt" }
 yktrace = { path = "../yktrace" }
-ykutil = { path = "../ykutil" }
 yksmp = { path = "../yksmp" }
 
 [features]

--- a/ykcapi/src/lib.rs
+++ b/ykcapi/src/lib.rs
@@ -16,7 +16,6 @@ use std::process;
 use std::{ptr, slice};
 use ykrt::{Location, MT};
 use yksmp::{Location as SMLocation, StackMapParser};
-use ykutil;
 
 // The "dummy control point" that is replaced in an LLVM pass.
 #[no_mangle]
@@ -56,16 +55,6 @@ pub extern "C" fn yk_location_new() -> Location {
 #[no_mangle]
 pub extern "C" fn yk_location_drop(loc: Location) {
     drop(loc)
-}
-
-/// Return a pointer to (and the size of) the .llvmbc section of the current executable.
-#[no_mangle]
-pub extern "C" fn __ykutil_get_llvmbc_section(res_addr: *mut *const c_void, res_size: *mut usize) {
-    let (addr, size) = ykutil::obj::llvmbc_section();
-    unsafe {
-        *res_addr = addr as *const c_void;
-        *res_size = size;
-    }
 }
 
 /// Reads out registers spilled to the stack of the previous frame during the deoptimisation

--- a/ykllvmwrap/src/lib.rs
+++ b/ykllvmwrap/src/lib.rs
@@ -16,5 +16,7 @@ extern "C" {
         faddr_keys: *const *const c_char,
         faddr_vals: *const *const c_void,
         len: size_t,
+        llvmbc_data: *const u8,
+        llvmbc_len: size_t,
     ) -> *const c_void;
 }

--- a/ykllvmwrap/src/lib.rs
+++ b/ykllvmwrap/src/lib.rs
@@ -12,10 +12,10 @@ extern "C" {
     pub fn __ykllvmwrap_irtrace_compile(
         func_names: *const *const c_char,
         bbs: *const size_t,
-        len: size_t,
+        trace_len: size_t,
         faddr_keys: *const *const c_char,
         faddr_vals: *const *const c_void,
-        len: size_t,
+        faddr_len: size_t,
         llvmbc_data: *const u8,
         llvmbc_len: size_t,
     ) -> *const c_void;

--- a/ykllvmwrap/src/ykllvmwrap.cc
+++ b/ykllvmwrap/src/ykllvmwrap.cc
@@ -34,6 +34,13 @@ using namespace llvm::orc;
 using namespace llvm::symbolize;
 using namespace std;
 
+struct BitcodeSection {
+  // Pointer to the start of the LLVM bitcode section.
+  void *data;
+  // The length of the LLVM bitcode section, in bytes.
+  size_t len;
+};
+
 // An annotator for `Module::print()` which adds debug location lines.
 class DebugAnnotationWriter : public AssemblyAnnotationWriter {
   string LastLineInfo;
@@ -54,8 +61,6 @@ public:
     }
   }
 };
-
-extern "C" void __ykutil_get_llvmbc_section(void **res_addr, size_t *res_size);
 
 // The bitcode module loaded from the .llvmbc section of the currently-running
 // binary. This cannot be shared across threads and used concurrently without
@@ -180,11 +185,8 @@ __yk_llvmwrap_symbolizer_find_code_sym(LLVMSymbolizer *Symbolizer,
 // Load the GlobalAOTMod.
 //
 // This must only be called from getAOTMod() for correct synchronisation.
-void loadAOTMod(void *Unused) {
-  void *SecPtr;
-  size_t SecSize;
-  __ykutil_get_llvmbc_section(&SecPtr, &SecSize);
-  auto Sf = StringRef((const char *)SecPtr, SecSize);
+void loadAOTMod(struct BitcodeSection &Bitcode) {
+  auto Sf = StringRef((const char *)Bitcode.data, Bitcode.len);
   auto Mb = MemoryBufferRef(Sf, "");
   SMDiagnostic Error;
   ThreadSafeContext AOTCtx = std::make_unique<LLVMContext>();
@@ -196,8 +198,8 @@ void loadAOTMod(void *Unused) {
 
 // Get a thread-safe handle on the LLVM module stored in the .llvmbc section of
 // the binary. The module is loaded if we haven't yet done so.
-ThreadSafeModule *getThreadAOTMod(void) {
-  std::call_once(GlobalAOTModLoaded, loadAOTMod, nullptr);
+ThreadSafeModule *getThreadAOTMod(struct BitcodeSection &Bitcode) {
+  std::call_once(GlobalAOTModLoaded, loadAOTMod, Bitcode);
   if (!ThreadAOTModInitialized) {
     ThreadAOTMod = cloneToNewContext(GlobalAOTMod);
     ThreadAOTModInitialized = true;
@@ -254,13 +256,13 @@ extern "C" void *compileModule(string TraceName, Module *M,
 // trace.
 //
 // Returns a pointer to the compiled function.
-extern "C" void *__ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[],
-                                              size_t Len, char *FAddrKeys[],
-                                              void *FAddrVals[],
-                                              size_t FAddrLen) {
+extern "C" void *__ykllvmwrap_irtrace_compile(
+    char *FuncNames[], size_t BBs[], size_t Len, char *FAddrKeys[],
+    void *FAddrVals[], size_t FAddrLen, void *BitcodeData, size_t BitcodeLen) {
   DebugIRPrinter DIP;
 
-  ThreadSafeModule *ThreadAOTMod = getThreadAOTMod();
+  struct BitcodeSection Bitcode = {BitcodeData, BitcodeLen};
+  ThreadSafeModule *ThreadAOTMod = getThreadAOTMod(Bitcode);
   // Getting the module without acquiring the context lock is safe in this
   // instance since ThreadAOTMod is not shared between threads.
   Module *AOTMod = ThreadAOTMod->getModuleUnlocked();

--- a/ykllvmwrap/src/ykllvmwrap.cc
+++ b/ykllvmwrap/src/ykllvmwrap.cc
@@ -257,7 +257,7 @@ extern "C" void *compileModule(string TraceName, Module *M,
 //
 // Returns a pointer to the compiled function.
 extern "C" void *__ykllvmwrap_irtrace_compile(
-    char *FuncNames[], size_t BBs[], size_t Len, char *FAddrKeys[],
+    char *FuncNames[], size_t BBs[], size_t TraceLen, char *FAddrKeys[],
     void *FAddrVals[], size_t FAddrLen, void *BitcodeData, size_t BitcodeLen) {
   DebugIRPrinter DIP;
 
@@ -272,8 +272,8 @@ extern "C" void *__ykllvmwrap_irtrace_compile(
   Module *JITMod;
   std::string TraceName;
   std::map<GlobalValue *, void *> GlobalMappings;
-  std::tie(JITMod, TraceName, GlobalMappings) =
-      createModule(AOTMod, FuncNames, BBs, Len, FAddrKeys, FAddrVals, FAddrLen);
+  std::tie(JITMod, TraceName, GlobalMappings) = createModule(
+      AOTMod, FuncNames, BBs, TraceLen, FAddrKeys, FAddrVals, FAddrLen);
   DIP.print(DebugIR::JITPreOpt, JITMod);
 #ifndef NDEBUG
   llvm::verifyModule(*JITMod, &llvm::errs());

--- a/yktrace/src/lib.rs
+++ b/yktrace/src/lib.rs
@@ -116,9 +116,9 @@ impl IRTrace {
     }
 
     pub fn compile(&self) -> *const c_void {
-        let len = self.len();
-        let mut func_names = Vec::with_capacity(len);
-        let mut bbs = Vec::with_capacity(len);
+        let trace_len = self.len();
+        let mut func_names = Vec::with_capacity(trace_len);
+        let mut bbs = Vec::with_capacity(trace_len);
         for blk in &self.blocks {
             if blk.is_unmappable() {
                 // The block was unmappable. Indicate this with a null function name.
@@ -143,7 +143,7 @@ impl IRTrace {
             ykllvmwrap::__ykllvmwrap_irtrace_compile(
                 func_names.as_ptr(),
                 bbs.as_ptr(),
-                len,
+                trace_len,
                 faddr_keys.as_ptr(),
                 faddr_vals.as_ptr(),
                 faddr_keys.len(),

--- a/yktrace/src/lib.rs
+++ b/yktrace/src/lib.rs
@@ -11,6 +11,7 @@ use std::{
     ptr,
 };
 mod hwt;
+use ykutil::obj::llvmbc_section;
 
 pub use errors::InvalidTraceError;
 pub use hwt::mapper::BlockMap;
@@ -136,6 +137,8 @@ impl IRTrace {
             faddr_vals.push(*k.1);
         }
 
+        let (llvmbc_data, llvmbc_len) = llvmbc_section();
+
         let ret = unsafe {
             ykllvmwrap::__ykllvmwrap_irtrace_compile(
                 func_names.as_ptr(),
@@ -144,6 +147,8 @@ impl IRTrace {
                 faddr_keys.as_ptr(),
                 faddr_vals.as_ptr(),
                 faddr_keys.len(),
+                llvmbc_data,
+                llvmbc_len,
             )
         };
         assert_ne!(ret, ptr::null());

